src/common/types/vector.cpp

```diff
@@ -291,17 +291,10 @@ void Vector::Resize(idx_t cur_size, idx_t new_size) {
 	}
 }
 
-// FIXME Just like DECIMAL, it's important that type_info gets considered when determining whether or not to cast
-// just comparing internal type is not always enough
-static bool ValueShouldBeCast(const LogicalType &incoming, const LogicalType &target) {
-	if (incoming.InternalType() != target.InternalType()) {
-		return true;
-	}
-	if (incoming.id() == LogicalTypeId::DECIMAL && incoming.id() == target.id()) {
-		//! Compare the type_info
-		return incoming != target;
-	}
-	return false;
+static bool ShouldCastValue(const LogicalType &left, const LogicalType &right) {
+	//! If two types share the same physical type and there is no extra type info
+	//  or the type info is the same, we can avoid casting
+	return (left.InternalType() != right.InternalType() || !left.EqualTypeInfo(right));
```

Comment:
Can't we just compare logical types here? Then we also don't need to export the `EqualTypeInfo` function.

Reply:
If we compare LogicalTypes then we will be casting in a lot of cases where this isn't necessary, because of the LogicalTypeIds.
This probably wouldn't matter too much because I believe SetValue isn't used where performance is crucial, but a lot of those casts are not implemented.

Reply:
I think casting by physical type is likely hiding bugs. We shouldn't be able to use `SetValue` to set an `INT32` value in a `DATE` vector, for example.

Reply:
Except we also do this intentionally in some places, like the HistogramFinalizeFunction, we store the values internally as their physical types, and then create a Value from them, and set that into the result vector.

I'm not really sure how to go about this, other than maybe a boolean flag on SetValue `no_cast` to show intent.